probe_src/probe_frontend/src/arena.rs

```diff
@@ -0,0 +1,370 @@
+    raw: Vec<u8>,
+    /// slice over Ops of the raw buffer.
+    ops: &'a [ffi::Op],
+}
+
+impl<'a> OpsArena<'a> {
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self> {
+        if bytes.len() < size_of::<ArenaHeader>() {
+            return Err(eyre!(
+                "Arena buffer too small, got {}, minimum size {}",
+                bytes.len(),
+                size_of::<ArenaHeader>()
+            ));
+        }
+
+        let header = unsafe { get_header_unchecked(&bytes) };
+        if header.capacity != bytes.len() {
+            return Err(eyre!(
+                "Invalid arena capacity, expected {}, got {}",
+                header.capacity,
+                bytes.len(),
+            ));
+        }
+        if header.used > header.capacity {
+            return Err(eyre!(
+                "Arena size {} is greater than capacity {}",
+                header.used,
+                header.capacity,
+            ));
+        }
+        if ((header.used - size_of::<ArenaHeader>()) % size_of::<ffi::Op>()) != 0 {
+            return Err(eyre!(
+                "Arena alignment error: used arena size minus header isn't a multiple of op size"
+            ));
+        }
+
+        let count = (header.used - size_of::<ArenaHeader>()) / size_of::<ffi::Op>();
+
+        log::debug!("[unsafe] converting Vec<u8> to &[ffi::Op]");
+        let ops = unsafe {
+            let ptr = bytes.as_ptr().add(size_of::<ArenaHeader>()) as *const ffi::Op;
+            std::slice::from_raw_parts(ptr, count)
+        };
+
+        Ok(Self { raw: bytes, ops })
+    }
+
+    pub fn decode(self, ctx: &ArenaContext) -> Result<Vec<ops::Op>> {
+        self.ops
+            .iter()
+            .map(|x| ops::Op::ffi_from(x, ctx))
+            .collect::<Result<Vec<_>>>()
+            .wrap_err("Failed to decode arena ops")
+    }
+}
+
+/// This struct represents a single `data/*.dat` arena allocator file emitted by libprobe.
+pub struct DataArena {
+    header: ArenaHeader,
+    raw: Vec<u8>,
+}
+
+impl DataArena {
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self> {
+        if bytes.len() < size_of::<ArenaHeader>() {
+            return Err(eyre!(
+                "Arena buffer too small, got {}, minimum size {}",
+                bytes.len(),
+                size_of::<ArenaHeader>()
+            ));
+        }
+        let header = unsafe { get_header_unchecked(&bytes) };
+        if header.capacity != bytes.len() {
+            return Err(eyre!(
+                "Invalid arena capacity, expected {}, got {}",
+                header.capacity,
+                bytes.len(),
+            ));
+        }
+        if header.used > header.capacity {
+            return Err(eyre!(
+                "Arena size {} is greater than capacity {}",
+                header.used,
+                header.capacity,
+            ));
+        }
+        Ok(Self { header, raw: bytes })
+    }
+
+    pub fn try_deref(&self, ptr: usize) -> Option<*const u8> {
```

Comment:
Can you list some pros/cons of returning a `*const u8` vs a `&[u8]` that goes from `ptr - base` to `used`?

Obviously the `Option<_>` is necessary in either case.

Btw, I'm pretty sure a allocation (e.g., a single null-terminated string) will never span *more* than one arena. If the allocation is larger than the entire arena, we increase the arena-size and allocate the request from a new arena.

Reply:
Assuming the header is properly generated, it would be memory safe to return a `&[u8]`. Part of me feels like a slice doesn't encapsulate the *idea* that we're just adjusting the pointer into the current virtual memory space. Although tbh the function name pretty bad, since at no point is the pointer actually being dereferenced. 

I think what I'll do is rename the function to `try_get_slice` and have it return an `Option<&[u8]>`. With a slice, it's also a bit easier to manipulate the bytes in rust, so I'll probably ditch the ffi call to `strlen()` since that's the only direct libc call in the entire codebase.

Reply:
Done in https://github.com/charmoniumQ/PROBE/pull/20/commits/70649ba0efd15a8ae10e0f257aeadf41a866fb43.